engine/googletest_result.cpp

```diff
@@ -0,0 +1,520 @@
+const optional< int >&
+engine::googletest_result::argument(void) const
+{
+    return _argument;
+}
+
+
+/// Gets the optional reason of the result.
+///
+/// \return The reason of the result if present; none otherwise.
+const optional< std::string >&
+engine::googletest_result::reason(void) const
+{
+    return _reason;
+}
+
+
+/// Checks whether the result should be reported as good or not.
+///
+/// \return True if the result can be considered "good", false otherwise.
+bool
+engine::googletest_result::good(void) const
+{
+    switch (_type) {
+    case googletest_result::disabled:
+    case googletest_result::skipped:
+    case googletest_result::successful:
+        return true;
+
+    case googletest_result::broken:
+    case googletest_result::failed:
+        return false;
+
+    default:
+        UNREACHABLE;
+    }
+}
+
+
+/// Reinterprets a raw result based on the termination status of the test case.
+///
+/// This reinterpretation ensures that the termination conditions of the program
+/// match what is expected of the paticular result reported by the test program.
+/// If such conditions do not match, the test program is considered bogus and is
+/// thus reported as broken.
+///
+/// This is just a helper function for calculate_result(); the real result of
+/// the test case cannot be inferred from apply() only.
+///
+/// \param status The exit status of the test program, or none if the test
+/// program timed out.
+///
+/// \result The adjusted result.  The original result is transformed into broken
+/// if the exit status of the program does not match our expectations.
+engine::googletest_result
+engine::googletest_result::apply(const optional< process::status >& status)
+    const
+{
+    if (!status) {
+        return *this;
+    }
+
+    INV(status);
+    switch (_type) {
+    case googletest_result::broken:
+        return *this;
+
+    case googletest_result::failed:
+        if (status.get().exited() && status.get().exitstatus() == EXIT_FAILURE)
+            return *this;
+        return googletest_result(googletest_result::broken,
+            "Failed test case should have reported failure but " +
+            format_status(status.get()));
+
+    case googletest_result::disabled:
+        if (status.get().exited() && status.get().exitstatus() == EXIT_SUCCESS)
+            return *this;
+        return googletest_result(googletest_result::broken,
+            "Disabled test case should have reported success but " +
+            format_status(status.get()));
+
+    case googletest_result::skipped:
+        if (status.get().exited() && status.get().exitstatus() == EXIT_SUCCESS)
+            return *this;
+        return googletest_result(googletest_result::broken,
+            "Skipped test case should have reported success but " +
+            format_status(status.get()));
+
+    case googletest_result::successful:
+        if (status.get().exited() && status.get().exitstatus() == EXIT_SUCCESS)
```

Comment:
All the branches in this switch statement check for the exact same exit code... which is suspicious. I don't think the behavior exposed here is tested.

Reply:
`broken` doesn't.

Reply:
Oh, I see what's missing. Please note that the failed case checks for the `exitstatus()` being `EXIT_FAILURE`.